Playwright: Run some real tests in Migration.E2E.Tests

HaoK opened the pull request:
Summary of changes
- Includes both the testassets/BasicApp contents, in addition to the published wwwroot merged into the helix payload 
- Add two real e2e tests to make sure things work and prototype the changes
- Changes to make dotnet test work now
- Refactor playwright test base from BlazorTemplates + this into a BrowserTestBase in the shared BrowserTesting library
- BrowserKind detected via first test method argument (defaults to Chrome otherwise), Chome/Firefox working, WebKit tbd

Issues:
- Firefox fails on linux due to profile directory issues (launcher might create/delete profile folder on each launch)
- ubuntu still skipped

Stability stats after moving to separate test project:
0 / 1 runs no build issues


Build issues seen sometimes:
```
.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.Razor\build\netstandard2.0\Microsoft.NET.Sdk.Razor.CodeGeneration.targets#L142
.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.Razor\build\netstandard2.0\Microsoft.NET.Sdk.Razor.CodeGeneration.targets(142,5): error MSB3231: (NETCORE_ENGINEERING_TELEMETRY=Build) Unable to remove directory "D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\Razor\". Could not find a part of the path '\\?\D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\Razor\ErrorBoundaryTest'.
```

```

##[error].dotnet\sdk\6.0.100-preview.5.21264.3\Microsoft.Common.CurrentVersion.targets(2202,5): error MSB3101: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not write state file "D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.csproj.AssemblyReference.cache". The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.csproj.AssemblyReference.cache' because it is being used by another process.
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Microsoft.Common.CurrentVersion.targets(2202,5): error MSB3101: Could not write state file "D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.csproj.AssemblyReference.cache". The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.csproj.AssemblyReference.cache' because it is being used by another process. [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
##[error].dotnet\sdk\6.0.100-preview.5.21264.3\Microsoft.Common.CurrentVersion.targets(2202,5): error MSB3101: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not write state file "D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.csproj.AssemblyReference.cache". The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.csproj.AssemblyReference.cache' because it is being used by another process.
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018: The "GenerateBlazorWebAssemblyBootJson" task failed unexpectedly. [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018: System.IO.IOException: The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\blazor.boot.json' because it is being used by another process. [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at System.IO.Strategies.FileStreamHelpers.ValidateFileHandle(SafeFileHandle fileHandle, String path, Boolean useAsyncIO) in System.Private.CoreLib.dll:token 0x6005b6c+0x25 [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at System.IO.Strategies.FileStreamHelpers.CreateFileOpenHandle(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b67+0x78 [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at System.IO.Strategies.WindowsFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005c26+0x61 [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b64+0x2c [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x600583e+0xa8 [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at System.IO.File.Create(String path) in System.IO.FileSystem.dll:token 0x60000f0+0x0 [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at Microsoft.NET.Sdk.BlazorWebAssembly.GenerateBlazorWebAssemblyBootJson.Execute() in Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.dll:token 0x6000062+0x0 [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() in Microsoft.Build.dll:token 0x60015da+0x3e [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) in Microsoft.Build.dll:token 0x600145a+0x809 [D:\workspace\_work\1\s\src\Components\test\testassets\BasicTestApp\BasicTestApp.csproj]
##[error].dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(300,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Build) The "GenerateBlazorWebAssemblyBootJson" task failed unexpectedly.
System.IO.IOException: The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\blazor.boot.json' because it is being used by another process.
   at System.IO.Strategies.FileStreamHelpers.ValidateFileHandle(SafeFileHandle fileHandle, String path, Boolean useAsyncIO) in System.Private.CoreLib.dll:token 0x6005b6c+0x25
   at System.IO.Strategies.FileStreamHelpers.CreateFileOpenHandle(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b67+0x78
   at System.IO.Strategies.WindowsFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005c26+0x61
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b64+0x2c
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x600583e+0xa8
   at System.IO.File.Create(String path) in System.IO.FileSystem.dll:token 0x60000f0+0x0
   at Microsoft.NET.Sdk.BlazorWebAssembly.GenerateBlazorWebAssemblyBootJson.Execute() in Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.dll:token 0x6000062+0x0
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() in Microsoft.Build.dll:token 0x60015da+0x3e
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) in Microsoft.Build.dll:token 0x600145a+0x809
  BasicTestApp -> D:\workspace\_work\1\s\artifacts\bin\BasicTestApp\Release\net6.0\BasicTestApp.dll

```

Microsoft.NET.Sdk.Razor\build\netstandard2.0\Microsoft.NET.Sdk.Razor.CodeGeneration.targets
```
.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.Razor\build\netstandard2.0\Microsoft.NET.Sdk.Razor.CodeGeneration.targets(142,5): error MSB3231: (NETCORE_ENGINEERING_TELEMETRY=Build) Unable to remove directory "D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\Razor\". Access to the path '\\?\D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\Razor\AuthTest' is denied.

.dotnet\sdk\6.0.100-preview.5.21264.3\Sdks\Microsoft.NET.Sdk.Razor\build\netstandard2.0\Microsoft.NET.Sdk.Razor.CodeGeneration.targets(142,5): error MSB3231: (NETCORE_ENGINEERING_TELEMETRY=Build) Unable to remove directory "D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\Razor\". Access to the path '\\?\D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\Razor\ServerReliability' is denied.
```

HaoK committed:
Test stuff

HaoK committed:
Fix tests

HaoK committed:
Cleanup

HaoK committed:
Make tests work, add bind test example

HaoK committed:
Try stuff

HaoK committed:
Publish to testassets for helix as well

HaoK commented on src/Components/test/E2ETestMigration/Microsoft.AspNetCore.Components.Migration.E2ETests.csproj:
```diff
@@ -31,6 +31,36 @@
+      Properties="TestTrimmedApps=true;PublishDir=$(MSBuildProjectDirectory)\$(OutputPath)trimmed\StandaloneApp\;"
+      Condition="'$(TestTrimmedApps)' == 'true'" />
+
+    <ProjectReference
```
review: do we need to publish more than just the basic app? for all tests to work?  if not some tests in the future will fail

HaoK committed:
Fix publish

HaoK commented on src/Components/test/E2ETestMigration/Infrastructure/ServerFixtures/BlazorWasmTestAppFixture.cs:
```diff
@@ -0,0 +1,108 @@
+{
+    public class BlazorWasmTestAppFixture<TProgram> : WebHostServerFixture
+    {
+        public readonly bool TestTrimmedApps = false;
```
TODO: deal with trimming variant

HaoK committed:
Update Microsoft.AspNetCore.Components.Migration.E2ETests.csproj

HaoK committed:
Update ci.yml

HaoK committed:
Get rid of trimmed apps for now

HaoK committed:
Move playwright stuff to BrowserTestBase

HaoK committed:
Refactor cleanup

HaoK committed:
Add input file tests to pw

HaoK committed:
Delete README.md

Commit:
Migrate DownloadFileFromAnchor test to Playwright (#32499)

* Migrate DownloadFileFromAnchor test to Playwright
* Unquarantine migrated test

HaoK committed:
Enable rest of input file tests

HaoK commented on src/Components/test/E2ETestMigration/Tests/InputFileTest.cs:
```diff
@@ -0,0 +1,260 @@
+                canvas.height = image.naturalHeight;
+                context.drawImage(image, 0, 0, image.naturalWidth, image.naturalHeight);
+
+                return canvas.toDataURL().split(',').pop();");
```
REVIEW: this is generating an illegal return statement runtime error from eval

HaoK commented on src/Components/test/E2ETestMigration/Tests/InputFileTest.cs:
```diff
@@ -0,0 +1,260 @@
+
+            foreach (var file in files)
+            {
+                var fileContainer = await _page.WaitForSelectorAsync($"[id='file-{file.Name}']");
```
this selector format seems to fail on non windows

HaoK replied:
```
[6.386s] [PlaywrightSharp.Transport.Connection] [Information] pw:channel:command {"id":14,"guid":"Frame@16b5c8146932b9329bd0645c8b7b562b","method":"setInputFiles","params":{"selector":"#input-file","files":[{"name":"e8fe5afe7774448f839a07805c8e646e\\923894c437da4472a339efc0f9537931.txt","mimeType":"text/plain","buffer":"VGhpcyBmaWxlIHdhcyB1cGxvYWRlZCB0byB0aGUgYnJvd3NlciBhbmQgcmVhZCBmcm9tIC5ORVQu"}],"isPage":true}
[6.387s] [PlaywrightSharp.Playwright] [Information] 2021-05-12T22:08:30.136Z pw:api waiting for selector "#input-file"
[6.390s] [PlaywrightSharp.Playwright] [Information] 2021-05-12T22:08:30.140Z pw:api   selector resolved to visible <input multiple _bl_2="" type="file" id="input-file"/>
[6.632s] [PlaywrightSharp.Transport.Connection] [Information] pw:channel:response {"id":14}
[6.633s] [PlaywrightSharp.Transport.Connection] [Information] pw:channel:command {"id":15,"guid":"Frame@16b5c8146932b9329bd0645c8b7b562b","method":"waitForSelector","params":{"selector":"[id=\u0027file-923894c437da4472a339efc0f9537931.txt\u0027]","isPage":true}
[6.634s] [PlaywrightSharp.Playwright] [Information] 2021-05-12T22:08:30.384Z pw:api waiting for selector "[id='file-923894c437da4472a339efc0f9537931.txt']" to be visible
[36.641s] [PlaywrightSharp.Transport.Connection] [Information] pw:channel:response {"id":15,"error":{"error":{"message":"Timeout 30000ms exceeded.\n=========================== logs ===========================\nwaiting for selector \"[id='file-923894c437da4472a339efc0f9537931.txt']\" to be visible\n============================================================\nNote: use DEBUG=pw:api environment variable to capture Playwright logs.","stack":"TimeoutError: Timeout 30000ms exceeded.\n=========================== logs ===========================\nwaiting for selector \"[id='file-923894c437da4472a339efc0f9537931.txt']\" to be visible\n============================================================\nNote: use DEBUG=pw:api environment variable to capture Playwright logs.\n    at ProgressController.run (/datadisks/disk1/work/BBBB09E8/w/BDC809FC/e/.playwright/unix/native/package/lib/server/progress.js:75:30)\n    at Frame.waitForSelector (/datadisks/disk1/work/BBBB09E8/w/BDC809FC/e/.playwright/unix/native/package/lib/server/frames.js:525:27)\n    at FrameDispatcher.waitForSelector (/datadisks/disk1/work/BBBB09E8/w/BDC809FC/e/.playwright/unix/native/package/lib/dispatchers/frameDispatcher.js:63:124)\n    at DispatcherConnection.dispatch (/datadisks/disk1/work/BBBB09E8/w/BDC809FC/e/.playwright/unix/native/package/lib/dispatchers/dispatcher.js:178:52)","name":"TimeoutError"}}
[36.643s] [Microsoft.AspNetCore.Components.E2ETest.Tests.InputFileTest] [Error] Test threw an exception.
System.TimeoutException: Timeout 30000ms exceeded.
=========================== logs ===========================
waiting for selector "[id='file-923894c437da4472a339efc0f9537931.txt']" to be visible
============================================================
Note: use DEBUG=pw:api environment variable to capture Playwright logs.
```

HaoK commented on src/Components/test/E2ETestMigration/Tests/InputFileTest.cs:
```diff
@@ -0,0 +1,260 @@
+
+        protected async Task ClearAndType(string selector, string value)
+        {
+            await _page.EvalOnSelectorAsync(selector, "e => e.Value = ''");
```
todo: add extension static class and move helpers

HaoK commented on src/Components/test/E2ETestMigration/Tests/InputFileTest.cs:
```diff
@@ -0,0 +1,260 @@
+            Assert.True(response.Ok, "Got: " + response.StatusText + "from: " + url);
+            Output.WriteLine("Loaded page");
+
+            await MountTestComponentAsync<InputFileComponent>(page);
```
TODO: Lift MountTestComponentAsync<T>/initialize core to ComponentTestBase and allow just specifying component to mount, need to double check if any tests do anything special

HaoK commented on src/Components/test/E2ETestMigration/Tests/InputFileTest.cs:
```diff
@@ -0,0 +1,260 @@
+            _tempDirectory = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"));
+            Directory.CreateDirectory(_tempDirectory);
+
+            //Navigate(ServerPathBase, noReload: _serverFixture.ExecutionMode == ExecutionMode.Client);
```
E2E tests try to navigate/avoid reloads, revisit later (optimization?)

HaoK commented on src/Components/test/E2ETestMigration/Tests/InputFileTest.cs:
```diff
@@ -0,0 +1,260 @@
+        {
+        }
+
+        private IPage _page;
```
Safe to lift/reuse a single _page to the base test class?

HaoK committed:
Fix javascript return error

HaoK commented on src/Components/test/E2ETestMigration/Tests/InputFileTest.cs:
```diff
@@ -0,0 +1,260 @@
+            await _page.SetInputFilesAsync("#input-file", new string[] { file1.Path, file2.Path });
+
+            // Validate that the proper exception is thrown
+            var exceptionMessage = await _page.QuerySelectorAsync("#exception-message");
```
review: exception message selector are not showing up, do these need to wait/for selector?

HaoK replied:
e.Value != e.value....grr casing with no errors

HaoK commented:
Might need to add some explicit ordering to the build pipelines, there's contention for writing the basictestapp now
```
CSC(0,0): error CS2012: (NETCORE_ENGINEERING_TELEMETRY=Build) Cannot open 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.dll' for writing -- 'The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.dll' because it is being used by another process.'
```

HaoK committed:
Fix casing argh

HaoK committed:
Try getting rid of the .

HaoK commented:
@pranavkm @captainsafia have you guys ever seen this error from publishing before?  I've seen it happen a few times before, and now its happened on back to back builds, I copied the trimming target from E2E tests but made it unconditional for helix to get the published BasicTestApp for the payload.  Any ideas what could cause this?

```
  Microsoft.AspNetCore.Components.Migration.E2ETests -> D:\workspace\_work\1\s\artifacts\bin\Microsoft.AspNetCore.Components.Migration.E2ETests\Release\net6.0\Microsoft.AspNetCore.Components.Migration.E2ETests.dll
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Compressing Blazor WebAssembly publish artifacts. This may take a while...
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Compressing Blazor WebAssembly publish artifacts. This may take a while...
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Compressing Blazor WebAssembly publish artifacts. This may take a while...
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Compressing Blazor WebAssembly publish artifacts. This may take a while...
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Compressing Blazor WebAssembly publish artifacts. This may take a while...
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Compressing Blazor WebAssembly publish artifacts. This may take a while...
##[error].dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(627,5): error MSB6006: (NETCORE_ENGINEERING_TELEMETRY=Build) "D:\workspace\_work\1\s\.dotnet" exited with code 1.
  Microsoft.AspNetCore.Components.WebAssembly.Server.Tests -> D:\workspace\_work\1\s\artifacts\bin\Microsoft.AspNetCore.Components.WebAssembly.Server.Tests\Release\net6.0\Microsoft.AspNetCore.Components.WebAssembly.Server.Tests.dll
  Microsoft.AspNetCore.Components.WebAssembly.Server.Tests -> D:\workspace\_work\1\s\artifacts\bin\Microsoft.AspNetCore.Components.WebAssembly.Server.Tests\Release\net6.0\publish\
```

HaoK commented:
@javiercn do you want to start taking a look at this PR?  Its got most of the initial shape done now minus some minor refactoring.  We have all of the InputFiles test, and a few other tests migrated as some initial examples.  Could probably move some more of the initialization boilerplate up somewhere, but I want to get your initial thoughts on things before I spend time cleaning up and polishing

HaoK committed:
Fix filecontent type

HaoK committed:
Fix path for non windows

rainersigwald commented on src/Components/test/E2ETestMigration/Microsoft.AspNetCore.Components.Migration.E2ETests.csproj:
```diff
@@ -24,13 +24,21 @@
+    <ProjectReference
+      Include="..\testassets\BasicTestApp\BasicTestApp.csproj"
+      Targets="Publish"
+      Properties="PublishDir=$(OutputPath)helixcontent\BasicTestApp\"
```
This conflicts with line 32 above; BasicTestApp.csproj will now be built twice with different global properties (once with nothing special, and once with this PublishDir) in parallel, which is a race condition.

HaoK replied:
Thanks, that makes perfect sense!  I'll try removing the top line, the main thing I was trying to accomplish was publish the app so we can copy the output into the helix payload.

HaoK replied:
So I copied this pattern from the other test project here (except it has a TestTrimmedApps = true condition): https://github.com/dotnet/aspnetcore/blob/main/src/Components/test/E2ETest/Microsoft.AspNetCore.Components.E2ETests.csproj#L58

Do we also have a race condition there?

HaoK replied:
Well this is a problem, we get compile/build errors if I remove the non publish target reference as there are references to types in the basic test app.  I'm not sure what the right fix is now?

HaoK replied:
Perhaps there a better way for me to force the BasicTestApp to publish than the ProjectReference, I just need it to happen before the PublishAssets target that copies the published output

HaoK committed:
Refactor/cleanup

HaoK committed:
Try publishing a different way

HaoK committed:
Fix publish

HaoK committed:
Update Microsoft.AspNetCore.Components.Migration.E2ETests.csproj

HaoK committed:
Update Microsoft.AspNetCore.Components.Migration.E2ETests.csproj

HaoK committed:
Update bindtest

HaoK committed:
Update ci.yml

HaoK committed:
Remove unused code

HaoK commented:
Still seeing build flakiness:

```
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(627,5): error MSB6006: "D:\workspace\_work\1\s\.dotnet" exited with code 1. [D:\workspace\_work\1\s\src\Components\test\testassets\GlobalizationWasmApp\GlobalizationWasmApp.csproj]
##[error].dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(627,5): error MSB6006: (NETCORE_ENGINEERING_TELEMETRY=Build) "D:\workspace\_work\1\s\.dotnet" exited with code 1.
```

HaoK committed:
Reenable bin log for helix

HaoK committed:
Add support for browserkind

HaoK committed:
Skip browsers that aren't enabled

HaoK committed:
Fix server test base

HaoK commented:
Another potential variant of build ordering error in GenerateBlazorWebAssemblyBootJson:

```
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018: The "GenerateBlazorWebAssemblyBootJson" task failed unexpectedly. [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018: System.IO.IOException: The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\StandaloneApp\Release\net6.0\blazor.publish.boot.json' because it is being used by another process. [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at System.IO.Strategies.FileStreamHelpers.ValidateFileHandle(SafeFileHandle fileHandle, String path, Boolean useAsyncIO) in System.Private.CoreLib.dll:token 0x6005b6a+0x25 [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at System.IO.Strategies.FileStreamHelpers.CreateFileOpenHandle(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b65+0x78 [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at System.IO.Strategies.WindowsFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005c24+0x61 [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b62+0x2c [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x600583c+0xa8 [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at System.IO.File.Create(String path) in System.IO.FileSystem.dll:token 0x60000f0+0x0 [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at Microsoft.NET.Sdk.BlazorWebAssembly.GenerateBlazorWebAssemblyBootJson.Execute() in Microsoft.NET.Sdk.BlazorWebAssembly.Tasks.dll:token 0x6000062+0x0 [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() in Microsoft.Build.dll:token 0x60015da+0x3e [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
D:\workspace\_work\1\s\.dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) in Microsoft.Build.dll:token 0x600145a+0x809 [D:\workspace\_work\1\s\src\Components\WebAssembly\testassets\StandaloneApp\StandaloneApp.csproj]
##[error].dotnet\sdk\6.0.100-preview.5.21230.2\Sdks\Microsoft.NET.Sdk.BlazorWebAssembly\targets\Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets(546,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Build) The "GenerateBlazorWebAssemblyBootJson" task failed unexpectedly.
System.IO.IOException: The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\StandaloneApp\Release\net6.0\blazor.publish.boot.json' because it is being used by another process.
   at System.IO.Strategies.FileStreamHelpers.ValidateFileHandle(SafeFileHandle fileHandle, String path, Boolean useAsyncIO) in System.Private.CoreLib.dll:token 0x6005b6a+0x25
   at System.IO.Strategies.FileStreamHelpers.CreateFileOpenHandle(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b65+0x78
   at System.IO.Strategies.WindowsFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x6005c24+0x61
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x6005b62+0x2c
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) in System.Private.CoreLib.dll:token 0x600583c+0xa8
   at System.IO.File.Create(String path) in System.IO.FileSystem.dll:token 0x60000f0+0x0
   at Microsoft.NET.Sdk.BlazorWebAssembly.GenerateBlazorWebAssemblyBootJson.Execute() in Microsoft.NET.Sdk.BlazorWebAssembly.
```

HaoK committed:
Skip firefox on linux

HaoK committed:
Fix compilation

HaoK committed:
Skip webkit on linux

HaoK committed:
Update playwrightSettings.linux.json

Update .azure/pipelines/ci.yml

Co-authored-by: Safia Abdalla <safia@microsoft.com>

Update helix-matrix.yml

Update helix-matrix.yml

Break test to verify browser/platform coverage

Debug avail browsers

Update Build.props

Update ComponentTestBase.cs

Don't delete temp directory if not set

Update ComponentTestBase.cs

Update playwrightSettings.linux.json

Update playwrightSettings.ci.json

Update BinaryHttpClientTest.cs

Update BinaryHttpClientTest.cs

Update BinaryHttpClientTest.cs

Try turning off build in parallel

Create copy of basic test app -> mega test app

Update playwrightSettings.json

Fix publish path

Add new test project to slns, skip mariner

Revert chromium change on ci.linux

Reenable parallel builds

Turn off parallel again

Turn off test parallelization

Update src/Components/test/testassets/MegaTestApp/MegaTestApp.csproj

Co-authored-by: Safia Abdalla <safia@microsoft.com>

Delete mega test app

Update Microsoft.AspNetCore.Components.Migration.E2ETests.csproj

Update AspNetCore.sln

Update AspNetCore.sln

Update Build.props

Update Components.slnf

Update BrowserTestBase.cs

Update PlaywrightFixture.cs

HaoK commented:
3rd variation of build contention

```
.dotnet\sdk\6.0.100-preview.5.21230.2\Microsoft.Common.CurrentVersion.targets(3161,5): error MSB3554: (NETCORE_ENGINEERING_TELEMETRY=Build) Cannot write to the output file "D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.Resources.resources". The process cannot access the file 'D:\workspace\_work\1\s\artifacts\obj\BasicTestApp\Release\net6.0\BasicTestApp.Resources.resources' because it is being used by another process.
```

captainsafia commented on .azure/pipelines/ci.yml:
```diff
@@ -751,7 +751,7 @@ stages:
                   /p:CrossgenOutput=false /p:ASPNETCORE_TEST_LOG_DIR=artifacts/log $(_InternalRuntimeDownloadArgs)
           displayName: Restore interop projects
-        - script: ./eng/build.cmd -ci -nobl -noBuildRepoTasks -noRestore -test -all -noBuildNative -projects eng\helix\helix.proj
+        - script: ./eng/build.cmd -ci -noBuildRepoTasks -noRestore -test -all -noBuildNative -projects eng\helix\helix.proj
```
```suggestion
        - script: ./eng/build.cmd -ci -nobl -noBuildRepoTasks -noRestore -test -all -noBuildNative -projects eng\helix\helix.proj
```

captainsafia commented on src/Components/test/E2ETestMigration/Infrastructure/ServerTestBase.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public void Navigate(string relativeUrl, bool noReload = false)
+        {
+            //Browser.Navigate(_serverFixture.RootUri, relativeUrl, noReload);
```
Do we need to replace this with a Playwright API?

HaoK replied:
I'm not sure exactly what the E2e tests were using this call for, as it sounds like it was a optimization for performance:

https://github.com/dotnet/aspnetcore/blob/0bb4cc8595e17c93ed3db13d7af61d716472338d/src/Components/test/E2ETest/Tests/BindTest.cs#L32

```
            // On WebAssembly, page reloads are expensive so skip if possible
```

I left this here for now as a reminder to encourage discussion like this, as I don't have enough context to know if this is something we should/need to keep doing

HaoK replied:
I sort of replaced the Navigate call with the MountUri property which get's used as part of InitializeCore, so basically every test does a page load, there's no playwright page instance reuse across tests here since I wanted to start from the most separate state, as opposed to trying to reuse anything to see if this helps from a stability standpoint

https://github.com/dotnet/aspnetcore/blob/328910d83b56127165e94cc6e3db67ea799d154c/src/Components/test/E2ETestMigration/Infrastructure/ComponentTestBase.cs#L64

javiercn replied:
In playwright world is fine if we create a new context every time, they are meant to be isolated from each other.

captainsafia commented on src/Components/test/E2ETestMigration/Tests/BindTest.cs:
```diff
@@ -0,0 +1,1367 @@
+        }
+
+        //[Fact]
+        //public void CanBindTextbox_InitiallyPopulated()
```
Are you wanting to convert these tests before merging the PR or?

HaoK replied:
No, I just want a small set of tests to start with, there's 12 right now.  I don't want to onboard too many tests before we are happy and settle on the overall pattern.  Its already a bit of a hassle to have to update so many tests when changing some of test patterns.

I had some open questions still about whether it was safe or a good idea to reuse the browser navigation page (TestPage property), I had to do some slightly sketchy stuff to make this work with the BrowserKind attribute.

https://github.com/dotnet/aspnetcore/blob/328910d83b56127165e94cc6e3db67ea799d154c/src/Components/test/E2ETestMigration/Infrastructure/ComponentTestBase.cs#L50

I also wanted to make sure we were happy with the existing MountTestComponent pattern before going too much farther.

https://github.com/dotnet/aspnetcore/blob/328910d83b56127165e94cc6e3db67ea799d154c/src/Components/test/E2ETestMigration/Infrastructure/ComponentTestBase.cs#L25

This is the first time I've seen how these test were setup, but now is a good time to consider if there's any improvements to make here.  I don't know have anything in mind in terms of changes or improvements, but I wanted to double check we are happy with this method of test selection

HaoK replied:
In terms of next steps, I think once we are mostly happy with the shape of  things, I'll probably try turning on a single test in every file to play minesweeper and see what explodes and find where the gaps still are (in a future PR)

javiercn commented on src/Components/test/E2ETestMigration/Infrastructure/ComponentTestBase.cs:
```diff
@@ -0,0 +1,81 @@
+                return;
+            }
+
+            TestBrowser = await BrowserManager.GetBrowserInstance(browserKind, BrowserContextInfo);
```
Is this launching a new browser every time?

HaoK replied:
Yes, I still have to get to https://github.com/dotnet/aspnetcore/issues/31111 which will tackle figuring out the best way to share fixtures/browser instances.  Seems like we could just have a static BrowserManager instead of a shared fixture too right?

javiercn reviewed (commented):
Looks great so far!

HaoK commented:
/azp run

HaoK commented on .azure/pipelines/helix-matrix.yml:
```diff
@@ -1,5 +1,9 @@
 # We only want to run full helix matrix on main
-pr: none
+pr:
```
revert

HaoK commented:
/azp run

HaoK commented:
/azp run

HaoK commented:
@dotnet/aspnet-build have you guys seen any errors like these before?  These look like build race conditions, not sure if they are related to this PR or not:

From windows testing job
```
CSC(0,0): error CS0009: (NETCORE_ENGINEERING_TELEMETRY=Build) Metadata file 'D:\workspace\_work\1\s\artifacts\bin\Microsoft.AspNetCore.Hosting.Server.Abstractions\Release\net6.0\Microsoft.AspNetCore.Hosting.Server.Abstractions.dll' could not be opened -- The process cannot access the file 'D:\workspace\_work\1\s\artifacts\bin\Microsoft.AspNetCore.Hosting.Server.Abstractions\Release\net6.0\Microsoft.AspNetCore.Hosting.Server.Abstractions.dll' because it is being used by another process.
```

From a helix job
```
.packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21278.1\tools\dotnet-cli\DotNetCli.targets(38,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) Unable to access dotnet cli runtime version 6.0.0-preview.6.21302.1 at https://dotnetcli.blob.core.windows.net/dotnet/Runtime/6.0.0-preview.6.21302.1/dotnet-runtime-6.0.0-preview.6.21302.1-win-x64.zip, An error occurred while sending the request.
```

wtgodbe commented:
The second looks like a legit networking error, I confirmed that the file exists at that location. As for the first, we see errors like that occasionally during build races, but I'm not sure there's a consistent cause. The binlog might give useful info.

HaoK commented:
Thanks @wtgodbe just wanted to make sure those aren't new from this PR, I will keep kicking builds

wtgodbe commented:
I wouldn't assume that the first error isn't new with this PR. The second one should be unrelated, though.

HaoK commented:
Do the azdo jobs all run on the same machine?  I ask because the test windows job shouldn't really be affected by this change unless its building in parallel on the same machine (seems like we can skip building all the helix ready test projects to cut down on parallelization if that is the case)

wtgodbe commented:
No, each job builds on a different machine

HaoK commented:
Cool, so its possible there's an existing race condition (from the components e2e test assembly maybe), but this PR shouldn't affect the windows test job

HaoK commented:
/azp run

SteveSandersonMS commented on src/Components/test/E2ETestMigration/Tests/BindTest.cs:
```diff
@@ -0,0 +1,1367 @@
+        [Theory]
+        [InlineData(BrowserKind.Chromium)]
+        [InlineData(BrowserKind.Firefox)]
+        [InlineData(BrowserKind.Webkit)]
```
Will we be repeating this block at the start of every test? Is there some way it could be made into an external concern?

HaoK replied:
The main thing we need is the permutations where we need a test invocation per browser kind.  We could turn each test into a for loop which does each browser test, but that seems like asking for more trouble.  Totally open to different ways to factor these if you have ideas.  The flow right now is the base class initializes the playwright page using reflection to look at the browser kind, so the tests can just work against TestPage.  

Do you favor something like?

```C#
[Fact]
public TestXyz() {
   foreach (var browser in SupportedTestBrowser) {
      var testPage = await GetTestPage(browser);
      // actually testXyz
   }
}
```

SteveSandersonMS commented on src/Components/test/E2ETestMigration/Tests/BindTest.cs:
```diff
@@ -0,0 +1,1367 @@
+        {
+            if (ShouldSkip(browserKind)) 
+            {
+                return;
```
I might be misreading this, but it looks as if any time we choose to skip a given test on a given browser, it would count as a "pass" as far as CI stats are concerned. If so that would be pretty misleading and might make it harder for us to reason about pass rate stats, whether to trigger retries, etc.

HaoK replied:
Yeah this is a bit funky, and worth discussing more, so the full problem is that, we have various helix queues for prs (win 10, ubuntu, osx), there's an even larger set for the full helix matrix that runs once a day.  And then there's 3 kinds of browsers (Chromium/Firefox/Webkit), there's a set of configuration files that specifies which browser and supported for which os.  Right now firefox is busted on ubuntu (so its skipped there, infrastructure related), so I was thinking that generally we will want to mark all tests with all 3 browsers, and this ShouldSkip check is what allows us to no-op the test if the browser is not supported on that OS.  The alternative to this would be getting rid of the BrowserKind attribute/parameter, and having a test per browser (would result in a different factoring of the tests, class or file per browserkind, but still doable if we want clean CI stats per test)

SteveSandersonMS commented on src/Components/test/E2ETestMigration/Tests/BindTest.cs:
```diff
@@ -0,0 +1,1367 @@
+
+            // Modify target; verify value is updated and that textboxes linked to the same data are updated
+            await target.TypeAsync("Changed value");
+            // Doesn't update until change event
```
This is interesting. How does Playwright know this event is what we want it to wait for?

Also reading [the docs on this API](https://playwrightsharp.dev/api/PlaywrightSharp.ElementHandle.html#PlaywrightSharp_ElementHandle_GetInnerTextAsync_System_Nullable_System_Int32__), it's not clear to me how we should know that's what it does. Could you clarify where we get this information?

HaoK replied:
This took some manual playing around with to get work, I initially tried the wrong method which triggered the change update, this api simulates typing in the textbox, so its not until you press tab or enter that the browser seems to trigger the change event, so this test types some characters and verifies that things don't change before, the tab key is sent using PressAsync which causes things to update

HaoK commented:
/azp run

HaoK commented:
@captainsafia our nemesis has returned, did something change recently again? Haven't seen this error in a while.  I gave up on trying to resolve the race conditions and have just cloned BasicTestApp -> MegaTestApp to make progress, so there shouldn't be as much build contention in this PR anymore

```
src\Components\test\testassets\MegaTestApp\FormsTest\InputFileComponent.razor(8,54): error RZ9991: (NETCORE_ENGINEERING_TELEMETRY=Build) The attribute names could not be inferred from bind attribute 'bind-value'. Bind attributes should be of the form 'bind' or 'bind-value' along with their corresponding optional parameters like 'bind-value:event', 'bind:format' etc.
```

captainsafia commented on src/Components/test/testassets/MegaTestApp/MegaTestApp.csproj:
```diff
@@ -0,0 +1,61 @@
+
+    <!-- Project supports more than one language -->
+    <BlazorWebAssemblyLoadAllGlobalizationData>true</BlazorWebAssemblyLoadAllGlobalizationData>
+    <_UseRazorSourceGenerator>false</_UseRazorSourceGenerator>
```
```suggestion
    <UseRazorSourceGenerator>false</UseRazorSourceGenerator>
```

This outta do the trick. We updated the property name in a recent SDK version.

HaoK committed:
Sync

HaoK committed:
Update ServerFixture.cs

HaoK committed:
Fix publish for basic test app

HaoK committed:
Fix headers

HaoK committed:
Update InputFileTest.cs

HaoK committed:
Update DownloadAnchorTest.cs

HaoK committed:
Remove navigate commented out code

HaoK committed:
Merge branch 'haok/pw3' of https://github.com/dotnet/aspnetcore into haok/pw3

HaoK committed:
Quarantine tests for now

captainsafia commented:
Huzzah!

HaoK commented:
> Huzzah!

I just binged Mythic Quest last week, so I can't help but think of Everlight when I hear that :)
